File: Codes/Hash/Non-Cryptographic/PJWHash/PJWHash.pas
{
    Hash function by Peter J. Weinberger
    Archive of Reversing.ID
    Non-Cryptographic Hash
}
function PJWHash(const key : String) : Cardinal;
const BitsInCardinal = Sizeof(Cardinal) * 8;
const ThreeQuarters  = (BitsInCardinal  * 3) div 4;
const OneEighth      = BitsInCardinal div 8;
const HighBits       : Cardinal = (not Cardinal(0)) shl (BitsInCardinal - OneEighth);
var
    i    : Cardinal;
    Test : Cardinal;
begin
    Result := 0;
    
    for i := 1 to Length(key) do
    begin
        Result := (Result shl OneEighth) + Ord(key[i]);
        Test   := Result and HighBits;
        If (Test <> 0) then
        begin
            Result := (Result xor (Test shr ThreeQuarters)) and (not HighBits);
        end;
    end;
end;

File: Codes/Hash/Non-Cryptographic/PJWHash/PJWHash.vb
'  Hash function by Peter J. Weinberger
'  Archive of Reversing.ID
'  Non-Cryptographic Hash
'

' Fungsi ini merupakan method dari sebuah class
Public Shared Function PJWFHash (key As String) As UInteger
    Const BitsInUnsignedInt As UInteger = CUInt(4 * 8)
	Const ThreeQuarters As UInteger = CUInt((BitsInUnsignedInt * 3) / 4)
	Const OneEighth As UInteger = CUInt(BitsInUnsignedInt \ 8)
	Const HighBits As UInteger = CUInt(&HFFFFFFFFUI) << CInt(BitsInUnsignedInt - OneEighth)
	Dim state As UInteger = 0
	Dim test As UInteger = 0
	Dim i As UInteger = 0

	For i = 0 To key.Length - 1
		state = (state << CInt(OneEighth)) + CByte(AscW(key(CInt(i))))
		test = state And HighBits

		If test <> 0 Then
			state = ((state Xor (test >> CInt(ThreeQuarters))) And (Not HighBits))
		End If
	Next

	Return state
End Function